### PR TITLE
*: implement index nested loop join

### PR DESCRIPTION
**Pull request by winoros**
When one of join's children is a scan and there are equal conditions that use the index of this scan. We can build this join to a nested loop join and build this scan using index to look up.
Currently index nested loop join is controlled by the hint `/*+ TIDB_INLJ(t1, t2) */` not cost-based.
And only the table in `TIDB_INLJ()` will be considered as outer table of the join.

### CODE REVIEW

**Comment by winoros** [comment]
PTAL @coocood @shenli @hanfei1991 @lamxTyler

**Comment by shenli** [comment]
@ilovesoup PTAL

File: plan/physical_plan_test.go [hunk 1]
```diff
@@ -1054,3 +1054,71 @@ func (s *testPlanSuite) TestScanController(c *C) {
...
+			sql: "select * from t t1 join t t2 on t1.a = t2.a",
+			ans: "LeftHashJoin{Table(t)->Table(t)}(t1.a,t2.a)",
+		},
+		{
+			sql: "select /*+ tidb_smj(t1, t2) */ * from t t1 join t t2 on t1.a = t2.a",
+			ans: "MergeJoin{Table(t)->Table(t)}(t1.a,t2.a)",
+		},
+		{
+			sql: "select /*+ tidb_inlj(t1, t2) */ * from t t1 join t t2 on t1.a = t2.a",
+			ans: "Apply{Table(t)->Selection->Table(t)->Cache}",
```
Comment (coocood) [inline plan/physical_plan_test.go]: Do we need the `Cache` plan?

File: executor/builder.go [hunk 2]
```diff
@@ -623,17 +623,30 @@ func (b *executorBuilder) buildSort(v *plan.Sort) Executor {
...
-	bigExec := b.build(v.Children()[0])
-	smallExec := b.build(v.Children()[1])
+	if v.SmallTable == 1 {
+		return &NestedLoopJoinExec{
+			SmallExec:     b.build(v.Children()[1]),
+			BigExec:       b.build(v.Children()[0]),
+			Ctx:           b.ctx,
+			BigFilter:     expression.ComposeCNFCondition(b.ctx, v.LeftConditions...),
+			SmallFilter:   expression.ComposeCNFCondition(b.ctx, v.RightConditions...),
+			OtherFilter:   expression.ComposeCNFCondition(b.ctx, append(expression.ScalarFuncs2Exprs(v.EqualConditions), v.OtherConditions...)...),
```
Comment (ilovesoup) [inline executor/builder.go]: I found a maybe bug here but not yet verify it:
When v.EqualConditions is prepared, its column index is based on both left / right child schema. But when used in otherFilter, it needs to be based on a merged schema. Column Index in scalar args need a re-resolve.

File: executor/builder.go [hunk 3]
```diff
@@ -623,17 +623,30 @@ func (b *executorBuilder) buildSort(v *plan.Sort) Executor {
...
-		OtherFilter: expression.ComposeCNFCondition(b.ctx, append(expression.ScalarFuncs2Exprs(v.EqualConditions), v.OtherConditions...)...),
-		schema:      v.Schema(),
-		outer:       v.JoinType != plan.InnerJoin,
+		SmallExec:     b.build(v.Children()[0]),
+		BigExec:       b.build(v.Children()[1]),
+		leftSmall:     true,
+		Ctx:           b.ctx,
+		BigFilter:     expression.ComposeCNFCondition(b.ctx, v.RightConditions...),
+		SmallFilter:   expression.ComposeCNFCondition(b.ctx, v.LeftConditions...),
+		OtherFilter:   expression.ComposeCNFCondition(b.ctx, append(expression.ScalarFuncs2Exprs(v.EqualConditions), v.OtherConditions...)...),
```
Comment (ilovesoup) [inline executor/builder.go]: Same here. Need a check.

File: plan/join_reorder.go [hunk 4]
```diff
@@ -27,7 +27,7 @@ func tryToGetJoinGroup(j *Join) ([]LogicalPlan, bool) {
 	// 1. already reordered
 	// 2. not inner join
 	// 3. forced merge join
```
Comment (ilovesoup) [inline plan/join_reorder.go]: Fix comment here.

File: parser/parser.y [hunk 5]
```diff
@@ -97,6 +97,7 @@ import (
 	describe		"DESCRIBE"
 	distinct		"DISTINCT"
 	tidbSMJ			"TIDB_SMJ"
+	tidbINLJ        "TIDB_INLJ"
```
Comment (shenli) [inline parser/parser.y]: Aligment

File: plan/join_reorder.go (hunk 4 above)
Comment (shenli) [inline plan/join_reorder.go]: Update the comment.

File: plan/physical_plan_builder.go [hunk 6]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
 	return resultInfo, nil
 }
 
+func convertCol2CorCol(cond expression.Expression, corCols []*expression.CorrelatedColumn, outerSchema *expression.Schema) expression.Expression {
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Add comment for this function.

File: plan/physical_plan_builder.go [hunk 7]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+	resultInfo.p = ap
+	if !allLeft {
+		resultInfo = enforceProperty(prop, resultInfo)
+	} else {
+		resultInfo = enforceProperty(limitProperty(prop.limit), resultInfo)
+	}
+	return resultInfo, nil
+}
+
+func (p *Join) convert2IndexNestedLoopJoinRight(prop *requiredProperty, innerJoin bool) (*physicalPlanInfo, error) {
```
Comment (ilovesoup) [inline plan/physical_plan_builder.go]: These two functions seem almost the same. Can we try make it one gracefully?
Very often ppl fix bug one place but ignored the copied version.

File: plan/physical_plan_builder.go [hunk 8]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+		if pos := outerSchema.ColumnIndex(x); pos >= 0 {
+			return corCols[pos]
+		}
+		return x
+	default:
+		return x
+	}
+}
+
+func (p *Join) buildSelectionWithConds(left bool) (*Selection, []*expression.CorrelatedColumn) {
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Add comment for this function.

File: plan/physical_plan_builder.go [hunk 9]
```diff
@@ -1475,6 +1740,7 @@ func addCachePlan(p PhysicalPlan, allocator *idAllocator) []*expression.Correlat
 		childCorCols := addCachePlan(child.(PhysicalPlan), allocator)
 		// If p is a Selection and controls the access condition of below scan plan, there shouldn't have a cache plan.
 		if sel, ok := p.(*Selection); len(selfCorCols) > 0 && len(childCorCols) == 0 && (!ok || !sel.ScanController) {
+			log.Warnf("child: %v", child.ID())
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: remove log

File: plan/stringer.go [hunk 10]
```diff
@@ -26,7 +26,7 @@ func ToString(p Plan) string {
 
 func toString(in Plan, strs []string, idxs []int) ([]string, []int) {
 	switch in.(type) {
-	case *Join, *Union, *PhysicalHashJoin, *PhysicalHashSemiJoin, *Apply, *PhysicalApply:
+	case *Join, *Union, *PhysicalHashJoin, *PhysicalHashSemiJoin, *Apply, *PhysicalApply, *PhysicalMergeJoin:
```
Comment (ilovesoup) [inline plan/stringer.go]: Thx.

File: plan/physical_plan_test.go (hunk 1 above)
Comment (winoros) [inline plan/physical_plan_test.go]: Don't need it. I will solve this soon.

File: plan/physical_plan_builder.go [hunk 11]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+		return nil, errors.Trace(err)
+	}
+	join := &PhysicalHashJoin{
+		LeftConditions: p.LeftConditions,
+		// TODO: decide concurrency by data size.
+		Concurrency:   JoinConcurrency,
+		DefaultValues: p.DefaultValues,
+		SmallTable:    1,
+	}
+	join.tp = "NestedLoopJoinLeft"
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Define a const.

File: plan/physical_plan_builder.go [hunk 12]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	join := &PhysicalHashJoin{
+		RightConditions: p.RightConditions,
+		// TODO: decide concurrency by data size.
+		Concurrency:   JoinConcurrency,
+		DefaultValues: p.DefaultValues,
+	}
+	join.tp = "NestedLoopJoinRight"
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Use a const.

File: executor/builder.go [hunk 13]
```diff
@@ -623,17 +623,30 @@ func (b *executorBuilder) buildSort(v *plan.Sort) Executor {
 }
 
 func (b *executorBuilder) buildNestedLoopJoin(v *plan.PhysicalHashJoin) *NestedLoopJoinExec {
-	bigExec := b.build(v.Children()[0])
-	smallExec := b.build(v.Children()[1])
+	if v.SmallTable == 1 {
```
Comment (shenli) [inline executor/builder.go]: Can we use a bool for SmallTable?

File: plan/physical_plan_builder.go (hunk 9 above)
Comment (shenli) [inline plan/physical_plan_builder.go]: This is a debug log?

File: plan/physical_plan_test.go [hunk 14]
```diff
@@ -1054,3 +1054,71 @@ func (s *testPlanSuite) TestScanController(c *C) {
...
+	cases := []struct {
+		sql string
+		ans string
+	}{
+		{
+			sql: "select * from t t1 join t t2 on t1.a = t2.a",
+			ans: "LeftHashJoin{Table(t)->Table(t)}(t1.a,t2.a)",
+		},
+		{
+			sql: "select /*+ tidb_smj(t1, t2) */ * from t t1 join t t2 on t1.a = t2.a",
```
Comment (shenli) [inline plan/physical_plan_test.go]: Add a test case that only contains one of the joined tables in hint list.

File: executor/join.go [hunk 15]
```diff
@@ -530,24 +532,33 @@ func (e *NestedLoopJoinExec) prepare() error {
...
-	newRow := &Row{
-		RowKeys: row.RowKeys,
-		Data:    make([]types.Datum, len(row.Data)+e.SmallExec.Schema().Len()),
+func (e *NestedLoopJoinExec) fillRowWithDefaultValue(bigRow *Row) (returnRow *Row) {
+	smallRow := &Row{
+		Data: make([]types.Datum, e.SmallExec.Schema().Len()),
 	}
-	copy(newRow.Data, row.Data)
-	return newRow
+	copy(smallRow.Data, e.defaultValues)
```
Comment (shenli) [inline executor/join.go]: This is a bug fix? Any test case cover this?

File: plan/join_reorder.go (hunk 4 above)
Comment (winoros) [inline plan/join_reorder.go]: added

File: plan/physical_plan_builder.go (hunk 11 above)
Comment (winoros) [inline plan/physical_plan_builder.go]: I found that many physical plan's tp is not a const, convert all of them to const?

File: executor/join.go (hunk 15 above)
Comment (winoros) [inline executor/join.go]: No, before this pr, only nested loop join with left as outer is used in apply, and default values seem to be nil always. So it is okay before.

File: plan/physical_plan_builder.go [hunk 16]
```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+func convertCol2CorCol(cond expression.Expression, corCols []*expression.CorrelatedColumn, outerSchema *expression.Schema) expression.Expression {
+	switch x := cond.(type) {
+	case *expression.ScalarFunction:
+		newArgs := make([]expression.Expression, 0, len(x.GetArgs()))
+		for _, arg := range x.GetArgs() {
+			newArg := convertCol2CorCol(arg, corCols, outerSchema)
+			newArgs = append(newArgs, newArg)
+		}
+		var newSf expression.Expression
+		if x.FuncName.L == ast.Cast {
```
Comment (winoros) [inline plan/physical_plan_builder.go]: NewFunction() could not new a cast function.

File: executor/builder.go (hunk 2 above)
Comment (winoros) [inline executor/builder.go]: add a resolve_indices in this builder

File: plan/join_reorder.go (hunk 4 above)
Comment (winoros) [inline plan/join_reorder.go]: done.

File: plan/physical_plan_test.go (hunk 14 above)
Comment (winoros) [inline plan/physical_plan_test.go]: addded

**Comment by winoros** [comment]
the comment of @coocood in previous pr #2834 is fixed.

**Comment by shenli** [comment]
@ilovesoup @hanfei1991 @coocood PTAL

File: plan/physical_plan_builder.go [hunk 17]
```diff
@@ -571,6 +572,230 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
 	return resultInfo, nil
 }
 
+// convertCol2CorCol will convert the column in the condition which can be found in outerSchema to a correlated column whose
+// Column is this column. And please make sure the outerSchema.Columns[i].Equal(corCols[i].Column)) holds when you call this.
+func convertCol2CorCol(cond expression.Expression, corCols []*expression.CorrelatedColumn, outerSchema *expression.Schema) expression.Expression {
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: move this to expression pkg.

File: plan/physical_plan_builder.go [hunk 18]
```diff
@@ -571,6 +572,230 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+	ap.tp = App
+	ap.allocator = p.allocator
+	ap.initIDAndContext(p.ctx)
+	ap.SetChildren(resultInfo.p.Children()...)
+	ap.SetSchema(resultInfo.p.Schema())
+	resultInfo.p = ap
+	if !allLeft {
+		resultInfo = enforceProperty(prop, resultInfo)
+	} else {
+		resultInfo = enforceProperty(limitProperty(prop.limit), resultInfo)
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: Why can we propagate limit ?

File: executor/join.go (hunk 15 above)
Comment (hanfei1991) [inline executor/join.go]: add default values test.

File: plan/logical_plan_builder.go [hunk 19]
```diff
@@ -231,6 +233,7 @@ func (b *planBuilder) buildJoin(join *ast.Join) LogicalPlan {
 
 	if b.TableHints() != nil {
 		joinPlan.preferMergeJoin = b.TableHints().ifPreferMergeJoin(leftAlias, rightAlias)
+		joinPlan.preferINLJ = b.TableHints().ifPreferINLJ(leftAlias, rightAlias)
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: If both preferMergeJoin and perferINLJ are true, we should report an error.

File: plan/physical_plan_builder.go [hunk 20]
```diff
@@ -571,6 +571,205 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
 	return resultInfo, nil
 }
 
+// buildSelectionWithConds will build a selection use the conditions of join and convert one side's column to correlated column.
+// This is called when build nested loop join.
+func (p *Join) buildSelectionWithConds(left bool) (*Selection, []*expression.CorrelatedColumn) {
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: `left` is not a good name.

File: plan/physical_plan_builder.go [hunk 21]
```diff
@@ -571,6 +571,205 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+		corCol.Column.ResolveIndices(outerSchema)
+		corCols = append(corCols, corCol)
+	}
+	selection := &Selection{baseLogicalPlan: newBaseLogicalPlan(Sel, p.allocator)}
+	selection.self = selection
+	selection.initIDAndContext(p.ctx)
+	selection.SetSchema(innerChild.Schema().Clone())
+	selection.SetChildren(innerChild)
+	if left {
+		conds = make([]expression.Expression, 0, len(p.EqualConditions)+len(p.RightConditions)+len(p.OtherConditions))
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: Do you need to resolve again for other condition?

File: plan/physical_plan_builder.go [hunk 22]
```diff
@@ -571,6 +571,205 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+		join.JoinType = InnerJoin
+	} else {
+		join.JoinType = LeftOuterJoin
+	}
+	join.SetSchema(p.Schema())
+	resultInfo := join.matchProperty(prop, lInfo, rInfo)
+	ap := &PhysicalApply{
+		PhysicalJoin: resultInfo.p,
+		OuterSchema:  corCols,
+	}
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: ap.self = ap

File: plan/physical_plan_builder.go [hunk 23]
```diff
@@ -571,6 +571,205 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
...
+		LeftConditions: p.LeftConditions,
+		// TODO: decide concurrency by data size.
+		Concurrency:   JoinConcurrency,
+		DefaultValues: p.DefaultValues,
+		SmallTable:    1,
+	}
+	join.tp = "NestedLoopJoinLeft"
+	join.allocator = p.allocator
+	join.initIDAndContext(p.ctx)
+	join.SetChildren(lInfo.p, rInfo.p)
```
Comment (hanfei1991) [inline plan/physical_plan_builder.go]: join.self = join

File: plan/physical_plan_test.go (hunk 1 above)
Comment (hanfei1991) [inline plan/physical_plan_test.go]: @winoros  you haven't fixed it.

**Comment by hanfei1991** [comment]
Can we offer a way to choose the outer table of index join ?

File: plan/physical_plan_builder.go (hunk 22 above)
Comment (winoros) [inline plan/physical_plan_builder.go]: But `self` is one part of baseLogicalPlan not one part of basePlan.

Comment (hanfei1991) [inline plan/physical_plan_builder.go]: well

**Comment by winoros** [comment]
Now we can use hint to choose the outer table.

**Comment by hanfei1991** [comment]
LGTM

File: executor/join.go (hunk 15 above)
Comment (winoros) [inline executor/join.go]: added

File: plan/physical_plan_test.go (hunk 1 above)
Comment (winoros) [inline plan/physical_plan_test.go]: This will be solved in other pr.

File: plan/physical_plan_builder.go (hunk 7 above)
Comment (winoros) [inline plan/physical_plan_builder.go]: This is the same with convert2HashJoinLeft and Right. I think it's okay.

**Comment by winoros** [comment]
PTAL @coocood @shenli

File: executor/join_test.go [hunk 24]
```diff
@@ -208,6 +208,20 @@ func (s *testSuite) TestJoin(c *C) {
...
+	tk.MustExec("create table t1(a int, b int)")
+	tk.MustExec("insert into t values(1, 1), (2, 2), (3, 3)")
+	tk.MustExec("insert into t1 values(1, 2), (1, 3), (3, 4), (4, 5)")
+	// The physical plans of the two sql are tested at physical_plan_test.go
+	tk.MustQuery("select /*+ TIDB_INLJ(t, t1) */ * from t join t1 on t.a=t1.a").Check(testkit.Rows("1 1 1 2", "1 1 1 3", "3 3 3 4"))
+	tk.MustQuery("select /*+ TIDB_INLJ(t1) */ * from t1 join t on t.a=t1.a and t.a < t1.b").Check(testkit.Rows("1 2 1 1", "1 3 1 1", "3 4 3 3"))
+	tk.MustQuery("select /*+ TIDB_INLJ(t, t1) */ * from t right outer join t1 on t.a=t1.a").Check(testkit.Rows("1 1 1 2", "1 1 1 3", "3 3 3 4", "<nil> <nil> 4 5"))
+	tk.MustQuery("select /*+ TIDB_INLJ(t, t1) */ avg(t.b) from t right outer join t1 on t.a=t1.a").Check(testkit.Rows("1.6667"))
+
+	_, err = tk.Exec("select /*+ TIDB_INLJ(t) TIDB_SMJ(t) * from t join t1 on t.a=t1.a")
```
Comment (shenli) [inline executor/join_test.go]: Syntax error.

File: plan/physical_plan_builder.go [hunk 25]
```diff
@@ -1004,7 +1246,7 @@ func (p *Union) convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanInfo,
 // makeScanController will try to build a selection that controls the below scan's filter condition,
 // and return a physicalPlanInfo. If the onlyJudge is true, it will only check whether this selection
 // can become a scan controller without building the physical plan.
-func (p *Selection) makeScanController(onlyJudge bool) (*physicalPlanInfo, bool) {
+func (p *Selection) makeScanController(onlyJudge bool) (*physicalPlanInfo, int) {
```
Comment (coocood) [inline plan/physical_plan_builder.go]: Replace `judge` with `check`.

File: plan/physical_plan_builder.go [hunk 26]
```diff
@@ -1041,79 +1283,88 @@ func (p *Selection) makeScanController(onlyJudge bool) (*physicalPlanInfo, bool)
...
+		}
+		// else will judge whether can build a index scan or just build a index scan.
+	}
+	var chosenPlan *PhysicalIndexScan
+	for _, expr := range p.Conditions {
+		if !expr.IsCorrelated() {
+			continue
 		}
-		ts := &PhysicalTableScan{
+		cond, _ := expression.SubstituteCorCol2Constant(expr)
```
Comment (coocood) [inline plan/physical_plan_builder.go]: Isn't the condition already been substituted at the beginning of the function?

File: plan/physical_plan_builder.go [hunk 27]
```diff
@@ -1033,87 +1275,89 @@ func (p *Selection) makeScanController(onlyJudge bool) (*physicalPlanInfo, bool)
...
+			newSel.SetChildren(child)
+			info := &physicalPlanInfo{
+				p:     &newSel,
+				count: uint64(ds.statisticTable.Count),
+			}
+			info.cost = float64(info.count) * selectionFactor
+			return info, controlTableScan
 		}
-		ts := &PhysicalTableScan{
+		// else will judge whether can build a index scan or just build a index scan.
```
Comment (coocood) [inline plan/physical_plan_builder.go]: Search the keyword `judge` replace them all, include comments.

File: plan/physical_plan_test.go [hunk 28]
```diff
@@ -1059,3 +1059,87 @@ func (s *testPlanSuite) TestScanController(c *C) {
...
+		is, err := mockResolve(stmt)
+		c.Assert(err, IsNil)
+
+		builder := &planBuilder{
+			allocator: new(idAllocator),
+			ctx:       mockContext(),
+			colMapper: make(map[*ast.ColumnNameExpr]int),
+			is:        is,
+		}
+		p := builder.build(stmt)
```
Comment (hanfei1991) [inline plan/physical_plan_test.go]: combine the line 1129 to 1142 to remove the `Cache` plan.

Comment (winoros) [inline plan/physical_plan_test.go]: Since line 1130 is a `c.Assert`, I do this from line 1131

File: plan/physical_plan_builder.go [hunk 29]
```diff
@@ -1271,41 +1271,40 @@ func (p *Selection) makeScanController() *physicalPlanInfo {
...
-			var accessConds []expression.Expression
-			accessConds, _, is.accessEqualCount, is.accessInAndEqCount = DetachIndexScanConditions(condsBackUp, is.Index)
-			if len(accessConds) == 0 {
-				continue
-			}
-			better := chosenPlan == nil || chosenPlan.accessEqualCount < is.accessEqualCount
-			better = better || (chosenPlan.accessEqualCount == is.accessEqualCount && chosenPlan.accessInAndEqCount < is.accessInAndEqCount)
+			_, _, accessEqualCount, accessInAndEqCount := DetachIndexScanConditions(condsBackUp, idx)
+			better := chosenPlan == nil || bestEqualCount < accessEqualCount
+			better = better || (bestEqualCount == accessEqualCount && bestInAndEqCount < accessInAndEqCount)
```
Comment (coocood) [inline plan/physical_plan_builder.go]: I think simply check access equal count is enough.
`better := accessEqualCount >= bestEqualCount`
Other wise, we need to add tests for `accesInAndEqCount`.

**Pull request merged**